### PR TITLE
Exclude export qbo ui

### PR DESCRIPTION
**Pull request by anishfyle**

### CODE REVIEW

File: src/app/core/models/configuration/advanced-setting.model.ts [hunk 1]
```diff
@@ -67,7 +68,8 @@ export class AdvancedSettingModel {
         auto_create_merchants_as_vendors: advancedSettingsForm.get('autoCreateMerchantsAsVendors')?.value,
         je_single_credit_line: advancedSettingsForm.get('singleCreditLineJE')?.value,
         change_accounting_period: advancedSettingsForm.get('changeAccountingPeriod')?.value,
-        memo_structure: advancedSettingsForm.get('memoStructure')?.value
+        memo_structure: advancedSettingsForm.get('memoStructure')?.value,
+        skip_export: advancedSettingsForm.get('skipExport')?.value
```
Comment (ashwin1111) [inline src/app/core/models/configuration/advanced-setting.model.ts:72]: We don't have any column in backend for this right, why do we need to send this?

File: src/app/core/models/configuration/skip-export-response.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { SkipExport } from "../misc/skip-export.model";
+
+export type ExpenseFilterResponse = {
```
Comment (ashwin1111) [inline src/app/core/models/configuration/skip-export-response.ts:3]: add this to skip export model, we're maintaining same model file for similar types in new app

File: src/app/core/models/misc/condition-field.model.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export type ConditionField = {
```
Comment (ashwin1111) [inline src/app/core/models/misc/condition-field.model.ts:1]: add to skip-export model

File: src/app/core/models/misc/skip-export.model.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+export type SkipExport = {
+  condition: string;
+  custom_field_type: any;
+  operator: 'isnull' | 'iexact' | 'icontains' | 'lt' | 'lte';
```
Comment (ashwin1111) [inline src/app/core/models/misc/skip-export.model.ts:4]: create enum for all these values

File: src/app/core/models/misc/skip-export.model.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+  values: string[];
+  rank: number;
+  join_by: 'AND' | 'OR' | null;
```
Comment (ashwin1111) [inline src/app/core/models/misc/skip-export.model.ts:7]: same

File: src/app/core/services/configuration/advanced-setting.service.ts [hunk 6]
```diff
@@ -31,6 +36,27 @@ export class AdvancedSettingService {
...
+    cacheBusterNotifier: skipExportCache
+  })
+  postSkipExport(workspaceId: number, skipExport: SkipExport): Observable<SkipExport> {
+    return this.apiService.post(`/workspaces/${this.workspaceService.getWorkspaceId()}/fyle/expense_filters/`, skipExport);
+  }
+
+  @CacheBuster({
+    cacheBusterNotifier: skipExportCache
+  })
+  deleteSkipExport(rank: number): Observable<SkipExport> {
```
Comment (ashwin1111) [inline src/app/core/services/configuration/advanced-setting.service.ts:56]: `deleteExpenseFilter()`

File: src/app/core/services/configuration/advanced-setting.service.ts [hunk 7]
```diff
@@ -31,6 +36,27 @@ export class AdvancedSettingService {
...
+    cacheBusterObserver: skipExportCache
+  })
+  getSkipExport(workspaceId: number): Observable<ExpenseFilterResponse> {
+    return this.apiService.get(`/workspaces/${this.workspaceService.getWorkspaceId()}/fyle/expense_filters/`, {});
+  }
+
+  @CacheBuster({
+    cacheBusterNotifier: skipExportCache
+  })
+  postSkipExport(workspaceId: number, skipExport: SkipExport): Observable<SkipExport> {
```
Comment (ashwin1111) [inline src/app/core/services/configuration/advanced-setting.service.ts:49]: `postExpenseFilter()`

File: src/app/core/services/configuration/advanced-setting.service.ts [hunk 8]
```diff
@@ -31,6 +36,27 @@ export class AdvancedSettingService {
     return this.apiService.put(`/v2/workspaces/${this.workspaceService.getWorkspaceId()}/advanced_configurations/`, exportSettingsPayload);
   }
 
+  @Cacheable({
+    cacheBusterObserver: skipExportCache
+  })
+  getSkipExport(workspaceId: number): Observable<ExpenseFilterResponse> {
```
Comment (ashwin1111) [inline src/app/core/services/configuration/advanced-setting.service.ts:42]: `getExpenseFilter()`

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html [hunk 9]
```diff
@@ -123,9 +123,182 @@ <h5 class="advanced-settings--memo-preview paragraph">
...
       </app-configuration-toggle-field>
 
+      <app-configuration-toggle-field *ngIf="!isLoading" [form]="advancedSettingsForm"
+        [iconPath]="'assets/images/pngs/vector-skip.svg'"
+        [label]="'Skip export of specific expenses from Fyle to QuickBooks Online'"
+        [subLabel]="'You could choose to skip the export of certain expenses from Fyle to QBO by setting up a conditional rule.'"
+        [formControllerName]="'skipExport'">
+      </app-configuration-toggle-field>
+      <div *ngIf="advancedSettingsForm.value.skipExport">
+       <div *ngIf="!isLoading" fxLayout="column" fxLayoutAlign="start stretch">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html:133]: can add the ngIf check to the previous line

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html [hunk 10]
```diff
@@ -123,9 +123,182 @@ <h5 class="advanced-settings--memo-preview paragraph">
...
+                <div fxFlex="15"></div>
+                <div class="fields-header-1" fxFlex="25">
+                   Condition
+                   <app-mandatory-field></app-mandatory-field>
+                </div>
+                <div class="fields-header-2" fxFlex="25">
+                   Operator
+                   <app-mandatory-field></app-mandatory-field>
+                </div>
+                <div class="fields-header-3" fxFlex="25" *ngIf = "(!skipExportForm.value.join_by) && ((skipExportForm.value.operator1 !== 'is_empty') && (skipExportForm.value.operator1 !== 'is_not_empty'))">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html:145]: move the ngIf logic to the ts, looks complicated to read

Comment (ashwin1111) [reply src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html:145]: add comments to the code too

Comment (ashwin1111) [reply src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html:145]: same for rest of the file, wherever the condition seems complicated, move them to ts

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html [hunk 11]
```diff
@@ -123,9 +123,182 @@ <h5 class="advanced-settings--memo-preview paragraph">
...
+                   <input matInput type="text" formControlName="value1" placeholder="Enter a value">
+                </mat-form-field>
+                <mat-form-field class="custom-mat-form-field" floatLabel="always" fxFlex="25" fxLayoutAlign="start stretch" appearance="outline" *ngIf="(skipExportForm.value.condition1?.field_name !== 'report_title') && (!skipExportForm.value.condition1 || skipExportForm.value.condition1.type==='SELECT' || skipExportForm.value?.condition1?.type==='TEXT' || skipExportForm.value?.condition1?.type==='NUMBER') && (skipExportForm.value.operator1 !== 'is_empty')  && (skipExportForm.value.operator1 !== 'is_not_empty')">
+                   <mat-chip-list #chipList aria-label="" [disabled]="isDisabledChip1">
+                   <mat-chip *ngFor="let item of valueOption1" [selectable]="true"
+                   [removable]="removable" (removed)="remove1(item)">
+                   {{item}}
+                   <img class="divider-icon" matChipRemove *ngIf="removable" src="../../../../../assets/images/pngs/vector-divider.svg">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.html:186]: use src like this `src="assets/../.."`

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.scss [hunk 12]
```diff
@@ -63,6 +63,247 @@
...
+}
+
+.error-message {
+  padding-top: 8px !important;
+}
+
+.skip-export-form2 {
+  margin-top: 2em;
+}
+// :host ::ng-deep .mat-form-field-appearance-outline .mat-form-field-flex {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.scss:213]: remove commented code

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 13]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
 
   private timeSpentEventRecorded: boolean = false;
 
+  // Skip Export
+  tooltip: boolean = false;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:90]: where is this used?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 14]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  skipFilterCount: number = 0;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:92]: name it `showExpenseFilters` and make it a boolean flag

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 15]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  skipExportForm: FormGroup;
+
+  showAdditionalCondition: boolean = false;
+
+  showAddButton: boolean = true;
+
+  workspaceId: number;
+
+  conditionFieldOptions: Array<{
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:110]: can use `: ConditionField[]` since we have a model for this

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 16]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+    field_name: string;
+    type: string;
+    is_custom: boolean;
+  }>;
+
+  valueFieldOptions1 = [];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:116]: where is this used?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 17]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  valueFieldOptions2 = [];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:118]: where is this used?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 18]
```diff
@@ -81,6 +86,43 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  operatorFieldOptions1: { label: string; value: string }[];
+
+  operatorFieldOptions2: { label: string; value: string }[];
+
+  joinByOptions = [{ value: 'AND' }, { value: 'OR' }];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:124]: use enum instead of hard coding

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 19]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
     this.windowReference = this.windowService.nativeWindow;
   }
 
+  visible = true;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:141]: where is this used?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 20]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  selectable = true;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:143]: not used anywhere

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 21]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  removable = true;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:145]: name it better, `isChipRemovable`

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 22]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  addOnBlur = true;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:147]: if we are not changing value for this, why not hard code as true in HTML?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 23]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+
+  customOperatorOptions = [
+    {
+      label: 'Is',
+      value: 'iexact'
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:152]: use enum

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 24]
```diff
@@ -96,6 +138,46 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+      label: 'Is not',
+      value: 'not_in'
+    }
+  ];
+
+  valueOption1: any[] = [];
+
+  valueOption2: any[] = [];
+
+  separatorKeysCodes: number[] = [ENTER, COMMA];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:179]: hard code in html

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 25]
```diff
@@ -126,10 +208,48 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+        this.skipExportForm.controls.join_by.setValue(null);
+        this.skipExportForm.controls.condition2.clearValidators();
+        this.skipExportForm.controls.operator2.clearValidators();
+        this.skipExportForm.controls.condition2.setValue(null);
+        this.skipExportForm.controls.operator2.setValue(null);
+        this.skipExportForm.controls.value2.clearValidators();
+        this.skipExportForm.controls.value2.setValue(null);
+      } else {
+        this.resetSkipExport();
+        this.skipExportForm.controls.condition1.clearValidators();
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:228]: can we reset whole form when the toggle is turned off?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 26]
```diff
@@ -126,10 +208,48 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+        this.skipExportForm.controls.value1.setValidators(Validators.required);
+        this.updateAdditionalFilterVisibility(false);
+        this.skipExportForm.controls.join_by.clearValidators();
+        this.skipExportForm.controls.join_by.setValue(null);
+        this.skipExportForm.controls.condition2.clearValidators();
+        this.skipExportForm.controls.operator2.clearValidators();
+        this.skipExportForm.controls.condition2.setValue(null);
+        this.skipExportForm.controls.operator2.setValue(null);
+        this.skipExportForm.controls.value2.clearValidators();
+        this.skipExportForm.controls.value2.setValue(null);
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:225]: for 2nd row, I don't think we need to clear validators and resetting values again since we already do it in the else case. Incase if this is needed, let me know the case for this

File: src/app/core/services/configuration/advanced-setting.service.ts (hunk 8 above)
Comment (ashwin1111) [reply src/app/core/services/configuration/advanced-setting.service.ts:42]: get workspace id from workspace service, don't receive it from component, for all methods

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 27]
```diff
@@ -202,18 +323,151 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+      this.mappingService.getFyleCustomFields(),
+      this.advancedSettingService.getSkipExport(this.workspaceId)
+
     ]).subscribe(response => {
       this.advancedSettings = response[0];
       this.billPaymentAccounts = response[1];
       this.workspaceGeneralSettings = response[2];
       this.adminEmails = this.advancedSettings.workspace_schedules?.additional_email_options ? this.advancedSettings.workspace_schedules?.additional_email_options.concat(response[3]) : response[3];
+
+      this.skipFilterCount = response[5].count;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:342]: do all these logic in a separate function

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 28]
```diff
@@ -242,7 +496,8 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
   }
 
   save(): void {
-    if (this.advancedSettingsForm.valid && !this.saveInProgress) {
+    this.saveSkipExportFields();
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:499]: this should be inside 500 block

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 29]
```diff
@@ -297,14 +552,355 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+    if (condition1.valid && condition2.valid && this.showAdditionalCondition) {
+      if (condition1.value?.field_name === condition2.value?.field_name) {
+          this.skipExportForm.controls.operator2.setValue(null);
+          return true;
+        }
+    }
+    return false;
+  }
+
+  saveSkipExportFields() {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:678]: move payload construction part to the model file

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 30]
```diff
@@ -297,14 +552,355 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
     });
   }
 
+  showTooltip() {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:555]: why is this needed?

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts [hunk 31]
```diff
@@ -297,14 +552,355 @@ export class AdvancedSettingsComponent implements OnInit, OnDestroy {
...
+    this.tooltip = true;
+  }
+
+  hideTooltip() {
+    this.tooltip = false;
+  }
+
+  add1(addEvent1: MatChipInputEvent): void {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:563]: have common function for add, remove etc

File: src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts (hunk 25 above)
Comment (anishfyle) [reply src/app/shared/components/configuration/advanced-settings/advanced-settings.component.ts:228]: this.skipExportForm.reset();
yes

**Review by ashwin1111 — APPROVED** [review]
